src/core/credGrainView.js

```diff
@@ -54,77 +54,137 @@ export type ParticipantsOptions = {|
 export class CredGrainView {
   _credGraph: CredGraph;
   _ledger: Ledger;
-  _defaultTimeScope: TimeScopedCredGrainView;
+  _participants: $ReadOnlyArray<ParticipantCredGrain>;
+  _intervals: IntervalSequence;
 
   constructor(credGraph: CredGraph, ledger: Ledger) {
     this._credGraph = credGraph;
     this._ledger = ledger;
-    this._defaultTimeScope = new TimeScopedCredGrainView(
-      credGraph,
-      ledger,
-      -Infinity,
-      Infinity
+    this._intervals = deepFreeze(credGraph.intervals());
+
+    const graphParticipants = new Map<IdentityId, GraphParticipant>();
+    for (const participant of credGraph.participants()) {
+      graphParticipants.set(participant.id, participant);
+    }
+
+    this._participants = deepFreeze(
+      ledger.accounts().map((account) => {
+        const graphParticipant = graphParticipants.get(account.identity.id);
+        if (!graphParticipant)
+          throw new Error(
+            `The graph is missing account [${account.identity.name}: ${account.identity.id}] that exists in the ledger.`
+          );
+
+        return {
+          identity: account.identity,
+          cred: graphParticipant.cred,
+          credPerInterval: graphParticipant.credPerInterval,
+          grainEarned: account.paid,
+          grainEarnedPerInterval: this._calculateGrainEarnedPerInterval(
+            account
+          ),
+        };
+      })
     );
   }
 
+  _calculateGrainEarnedPerInterval(account: Account): $ReadOnlyArray<?Grain> {
+    return this._intervals.map((interval) => {
+      let grain = G.ZERO;
+      account.allocationHistory.forEach((allocationReceipt) => {
+        if (
+          interval.startTimeMs < allocationReceipt.credTimestampMs &&
+          allocationReceipt.credTimestampMs <= interval.endTimeMs
+        )
+          grain = G.add(grain, allocationReceipt.grainReceipt.amount);
+      });
+      return grain;
+    });
```

Comment:
Let `A` be number of allocations and `I` be number of intervals. Then this runs in `O(A*I)` time but could run in `O(I+A)` time if we leverage the fact that allocation history is in time-sorted order. Ie., we don't actually need to iterate over every allocation for every interval.

Since we run this for every participant (`P`), we wind up with something like `O(P*I*A)` instead of `O(P * (I+A))`, inside the constructor. Having expensive constructors is kind of an anti-pattern already, so I think it might be worth pre-emptively optimizing this. It should be pretty straightforward to do so.

I'm going to approve this because I think the code is correct as written and we can optimize it in another PR (and maybe add some safety checks, like runtime verifying that the allocations are, in fact, in chronological order).

Reply:
I thought about that too, but wasn't sure if it was a worthwhile optimization. Sounds like yes, as a fast follow.